src/basic/fileio.c

```diff
@@ -403,18 +403,20 @@ int read_full_virtual_file(const char *filename, char **ret_contents, size_t *re
                         return -EBADF;
 
                 /* Be prepared for files from /proc which generally report a file size of 0. */
+                assert_cc(READ_FULL_BYTES_MAX < SSIZE_MAX);
                 if (st.st_size > 0) {
-                        if (st.st_size > SSIZE_MAX) /* safety check in case off_t is 64bit and size_t 32bit */
+                        if (st.st_size > READ_FULL_BYTES_MAX)
                                 return -E2BIG;
 
                         size = st.st_size;
                         n_retries--;
-                } else
-                        /* Double the buffer size (saturate in case of overflow) */
-                        size = size > SSIZE_MAX / 2 ? SSIZE_MAX : size * 2;
+                } else {
+                        /* Double the buffer size */
+                        if (size > READ_FULL_BYTES_MAX / 2)
+                                return -E2BIG;
```

Comment:
this means we'll in average never read more than 3/4th of READ_FULL_BYTES_MAX though?

mayb:
```c
if (size >= READ_FULL_BYTES_MAX)
        return -E2BIG;
if (size > READ_FULL_BYTE_MAX/2)
        size = READ_FULL_BYTES_MAX; /* clamp to max */
else
        size *= 2;

Reply:
No, I don't see this. We multiply by 2, so in principle we can hit `size==READ_FULL_BYTES_MAX`.

Reply:
But if its already READ_FULL_BYTES_MAX/2+1  then well give up  and not try again even if we could still almost double the size without actually hitting the limot

Reply:
Indeed. I updated the patch as suggested.
I now changed READ_FULL_BYTES_MAX to be 4MB-1, so that it's again the "maximum size of the file", so that we allocate 4MB evenly.